src/Eventso.Subscription.Kafka/DeadLetter/PoisonEventQueue.cs

```diff
@@ -0,0 +1,166 @@
+            
+        }
+    }
+
+    public void Revoke(TopicPartition topicPartition)
+    {
+        // no concurrency, same thread as consume
+        if (_partitionPoisonKeys.TryRemove(topicPartition, out var keysTask))
+        {
+            keysTask.CancellationTokenSource.Cancel();
+            keysTask.CancellationTokenSource.Dispose();
+
+            if (keysTask.Task.IsFaulted)
+                _ = keysTask.Task.Exception;
+        }
+
+        logger.PartitionRevoke(groupId, topicPartition.Topic, topicPartition.Partition);
+    }
+
+    public async ValueTask<bool> Contains(TopicPartition topicPartition, Guid key, CancellationToken token)
+    {
+        var keys = await GetTopicPartitionKeys(topicPartition, token);
+        return keys.Contains(key);
+    }
+
+    public async Task Enqueue(ConsumeResult<byte[], byte[]> @event, DateTime failureTimestamp, string failureReason, CancellationToken token)
+    {
+        var key = DeserializeKey(@event.Topic, @event.Message.Headers, @event.Message.Key);
+        logger.Enqueue(
+            groupId,
+            @event.TopicPartitionOffset.Topic,
+            @event.TopicPartitionOffset.Partition,
+            @event.TopicPartitionOffset.Offset,
+            key,
+            failureReason);
+        
+        var alreadyPoisoned = await poisonEventStore.CountPoisonedEvents(groupId, @event.TopicPartitionOffset.Topic, token);
+        if (alreadyPoisoned >= maxNumberOfPoisonedEventsInTopic)
+            throw new EventHandlingException(
+                @event.TopicPartitionOffset.Topic,
+                $"Dead letter queue size ({alreadyPoisoned}) exceeds threshold ({maxNumberOfPoisonedEventsInTopic}). " +
+                $"Poison event: {@event.TopicPartitionOffset}. " +
+                $"Error: {failureReason}",
+                null);
+        
+        await poisonEventStore.AddEvent(groupId, @event, failureTimestamp, failureReason, token);
+        
+        var keys = await GetTopicPartitionKeys(@event.TopicPartitionOffset.TopicPartition, token);
+        keys.Add(key);
+    }
+
+    public async IAsyncEnumerable<ConsumeResult<byte[], byte[]>> Peek([EnumeratorCancellation] CancellationToken token)
+    {
+        bool needToProcess = true;
+
+        while (needToProcess)
+        {
+            needToProcess = false;
+            var topicPartitions = _partitionPoisonKeys.Keys.ToArray();
+            foreach (var topicPartition in topicPartitions)
+            {
+                try
+                {
+                    var keys = await GetTopicPartitionKeys(topicPartition, token);
+                    if (keys.Count == 0)
+                        continue;
+                }
+                catch (Exception)
+                {
+                    continue;
+                }
+
+                var eventForRetrying = await retryScheduler.GetNextRetryTarget(groupId, topicPartition, token);
+                if (eventForRetrying == null)
+                    continue;
```

Comment:
too much control flow within nested loops, hard to reason about. it begs for refactoring